clients/drcachesim/drpt2trace/pt2ir.h

```diff
@@ -0,0 +1,272 @@
+         * trace time, and the sideband recodes time.
+         *
+         *   time_shift = perf_event_mmap_page.ttime_shift
+         *   time_mult = perf_event_mmap_page.time_mult
+         *   time_zero = perf_event_mmap_page.time_zero
+         */
+        uint16_t time_shift;
+        uint32_t time_mult;
+        uint64_t time_zero;
+
+        /* If the trace contains coarse timing information, the sideband events may be
+         * applied too late to cause "no memory mmapped at this address" issue. To set the
+         * tsc_offset can fix the issue.
+         * XXX: The libipt documentation said we need to set a suitable value for
+         * tsc_offset. But it didn't give a method to get a suitable value. So we may need
+         * to find a way to generate the suitable tsc_offset.
+         */
+        uint64_t tsc_offset;
+    } sb_config; /**< The libipt-sb config of PT raw trace. We can get these parameters by
+                  *   running libipt/scirpts/perf-get-opts.bash.
+                  */
+
+    /* sb_primary_file_path is the path of the primary sideband file.
+     * sb_secondary_file_path_list is a list of paths to secondary sideband files.
+     * The sideband primary file and secondary files are stored perf event records. A
+     * primary sideband file contain perf event records for the traced cpu.  A secondary
+     * sideband contain perf event records for other cpus on the system.
+     */
+    std::string sb_primary_file_path; /**< The sideband primary file of PT raw trace. */
+    std::vector<std::string>
+        sb_secondary_file_path_list; /**< The sideband secondary file of PT raw trace. */
+
+    std::string kcore_path; /**< The path of the kernel dump file. */
+};
+
+struct pt_image_section_cache;
+struct pt_sb_pevent_config;
+struct pt_sb_session;
+struct pt_insn_decoder;
+struct pt_ins;
+
+/**
+ * pt2ir_t is a class that can convert PT raw trace to DynamoRIO's IR.
+ */
+class pt2ir_t {
+public:
+    pt2ir_t();
+    ~pt2ir_t();
+
+    /**
+     * Returns TRUE if the pt2ir_t is successfully initialized. Returns FALSE on failure.
+     * \note Parse struct pt2ir_config_t and initialize PT instruction decoder, the
+     * sideband session, and images caches.
+     */
+    bool
+    init(IN const pt2ir_config_t pt2ir_config);
+
+    /**
+     * Returns pt2ir_convert_status_t.
+     * \note The convert function performs two processes: (1) decode the PT raw trace into
+     * libipt's IR format pt_insn; (2) convert pt_insn into the doctor's IR format
+     * instr_t. If the convertion is successful, the function returns PT2IR_CONV_SUCCESS.
+     * Otherwise, the function returns the corresponding error code.
+     */
+    pt2ir_convert_status_t
+    convert();
+
+    /**
+     * Returns all converted DynamoRIO IR.
+     */
+    std::vector<instr_t>
+    get_instrlist();
+
+    /**
+     * Returns the number of instructions in the converted IR.
+     */
+    uint64_t
+    get_instr_count();
```

Comment:
I thought get_instrlist will be implemented later, but if it is available already, then we don't need get_instr_count anymore. The caller can just print get_instrlist().size() themselves.